les/utils/limiter.go

```diff
@@ -0,0 +1,405 @@
+		if len(l.nodes) == 0 {
+			l.cond.Signal()
+		}
+		l.nodes[id] = nq
+		if nq.valueWeight != 0 {
+			l.valueSelect.Update(nq)
+		}
+		l.addToGroup(nq, address)
+	}
+	l.totalPriorWeight += priorWeight
+	if l.totalPriorWeight > l.totalPriorLimit {
+		l.dropRequests()
+	}
+	return process
+}
+
+// update updates the selection weights of the node queue
+func (l *Limiter) update(nq *nodeQueue) {
+	flatWeight, valueWeight := selectionWeights(nq.relCost, nq.value)
+	ag := l.addresses[nq.address]
+	ag.update(nq, flatWeight)
+	l.addressSelect.Update(ag)
+	if valueWeight != 0 {
+		nq.valueWeight = valueWeight
+		l.valueSelect.Update(nq)
+	}
+}
+
+// addToGroup adds the node queue to the given address group. The group is created if
+// it does not exist yet.
+func (l *Limiter) addToGroup(nq *nodeQueue, address string) {
+	nq.address = address
+	ag := l.addresses[address]
+	if ag == nil {
+		ag = &addressGroup{}
+		l.addresses[address] = ag
+	}
+	ag.add(nq)
+	l.addressSelect.Update(ag)
+}
+
+// removeFromGroup removes the node queue from its address group
+func (l *Limiter) removeFromGroup(nq *nodeQueue) {
+	ag := l.addresses[nq.address]
+	ag.remove(nq)
+	if len(ag.nodes) == 0 {
+		delete(l.addresses, nq.address)
+	}
+	l.addressSelect.Update(ag)
+}
+
+// remove removes the node queue from its address group, the nodes map and the value
+// selector
+func (l *Limiter) remove(nq *nodeQueue) {
+	l.removeFromGroup(nq)
+	if nq.valueWeight != 0 {
+		l.valueSelect.Remove(nq)
+	}
+	delete(l.nodes, nq.id)
+}
+
+// choose selects the next node queue to process.
+// Note: when a node queue becomes empty it stays in the random selectors for one more
+// selection round before removed, with a weight based on the last relative cost.
+// If no more requests are added to the queue before it is selected again then the
+// queue is removed and the next time a request comes from the same node the queue is
+// added with the highest possible weight.
+func (l *Limiter) choose() *nodeQueue {
+	if l.valueSelect.IsEmpty() || l.lastValueSelected {
+		if ag, ok := l.addressSelect.Choose().(*addressGroup); ok {
+			l.lastValueSelected = false
+			return ag.choose()
+		}
+	}
+	nq, _ := l.valueSelect.Choose().(*nodeQueue)
+	l.lastValueSelected = true
+	return nq
+}
+
+// processLoop processes requests sequentially
+func (l *Limiter) processLoop() {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	for {
+		if l.quit {
+			for _, nq := range l.nodes {
+				for _, request := range nq.queue {
+					close(request.process)
+				}
+			}
+			return
+		}
+		nq := l.choose()
+		if nq == nil {
+			l.cond.Wait()
+			continue
+		}
+		if len(nq.queue) > 0 {
+			request := nq.queue[0]
+			nq.queue = nq.queue[1:]
+			nq.totalPriorWeight -= request.priorWeight
+			l.totalPriorWeight -= request.priorWeight
+			nq.relCost = 0
+			l.lock.Unlock()
+			costCh := make(chan float64)
+			request.process <- costCh
+			relCost := <-costCh / float64(l.maxPriorWeight)
+			l.lock.Lock()
+			nq.relCost += relCost
```

Comment:
I'm not quite sure, maybe we can use assignment? `nq.relCost = relCost`? It's always reset before the unlock....